### ISSUE TITLE
Build failure with 5.4.78 kernel

### ISSUE THREAD

**Issue by ncopa** [sipwise/rtpengine#1111]
Build fails on x86_64 with 5.4.78 kernel due to https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-5.4.y&id=56907fa27b9496609cdc90485555a176a7d4c16b
```
make[1]: Entering directory '/usr/src/linux-headers-5.4.78-0-lts'
  CC [M]  /home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module/xt_RTPENGINE.o
/home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module/xt_RTPENGINE.c: In function 'send_proxy_packet4':
/home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module/xt_RTPENGINE.c:3355:42: error: passing argument 2 of 'ip_route_me_harder' from incompatible pointer type [-Werror=incompatible-pointer-types]
 3355 |  if (ip_route_me_harder(par->state->net, skb, RTN_UNSPEC))
      |                                          ^~~
      |                                          |
      |                                          struct sk_buff *
In file included from ./include/uapi/linux/netfilter_ipv4/ip_tables.h:22,
                 from ./include/linux/netfilter_ipv4/ip_tables.h:23,
                 from /home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module/xt_RTPENGINE.c:25:
./include/linux/netfilter_ipv4.h:19:54: note: expected 'struct sock *' but argument is of type 'struct sk_buff *'
   19 | int ip_route_me_harder(struct net *net, struct sock *sk, struct sk_buff *skb, unsigned addr_type);
      |                                         ~~~~~~~~~~~~~^~
/home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module/xt_RTPENGINE.c:3355:6: error: too few arguments to function 'ip_route_me_harder'
 3355 |  if (ip_route_me_harder(par->state->net, skb, RTN_UNSPEC))
      |      ^~~~~~~~~~~~~~~~~~
In file included from ./include/uapi/linux/netfilter_ipv4/ip_tables.h:22,
                 from ./include/linux/netfilter_ipv4/ip_tables.h:23,
                 from /home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module/xt_RTPENGINE.c:25:
./include/linux/netfilter_ipv4.h:19:5: note: declared here
   19 | int ip_route_me_harder(struct net *net, struct sock *sk, struct sk_buff *skb, unsigned addr_type);
      |     ^~~~~~~~~~~~~~~~~~
/home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module/xt_RTPENGINE.c: In function 'send_proxy_packet6':
/home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module/xt_RTPENGINE.c:3451:43: error: passing argument 2 of 'ip6_route_me_harder' from incompatible pointer type [-Werror=incompatible-pointer-types]
 3451 |  if (ip6_route_me_harder(par->state->net, skb))
      |                                           ^~~
      |                                           |
      |                                           struct sk_buff *
In file included from /home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module/xt_RTPENGINE.c:27:
./include/linux/netfilter_ipv6.h:146:55: note: expected 'struct sock *' but argument is of type 'struct sk_buff *'
  146 | int ip6_route_me_harder(struct net *net, struct sock *sk, struct sk_buff *skb);
      |                                          ~~~~~~~~~~~~~^~
/home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module/xt_RTPENGINE.c:3451:6: error: too few arguments to function 'ip6_route_me_harder'
 3451 |  if (ip6_route_me_harder(par->state->net, skb))
      |      ^~~~~~~~~~~~~~~~~~~
In file included from /home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module/xt_RTPENGINE.c:27:
./include/linux/netfilter_ipv6.h:146:5: note: declared here
  146 | int ip6_route_me_harder(struct net *net, struct sock *sk, struct sk_buff *skb);
      |     ^~~~~~~~~~~~~~~~~~~
cc1: some warnings being treated as errors
make[2]: *** [scripts/Makefile.build:266: /home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module/xt_RTPENGINE.o] Error 1
make[1]: *** [Makefile:1729: /home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-5.4.78-0-lts'
make: *** [Makefile:34: modules] Error 2
make: Leaving directory '/home/ncopa/aports/testing/rtpengine-lts/src/lts/kernel-module'
```

**Comment by ncopa** [comment]
I wonder if this is correct?
```diff
diff --git a/kernel-module/xt_RTPENGINE.c b/kernel-module/xt_RTPENGINE.c
index c12edcd..c5aacc7 100644
--- a/kernel-module/xt_RTPENGINE.c
+++ b/kernel-module/xt_RTPENGINE.c
@@ -3351,7 +3351,9 @@ static int send_proxy_packet4(struct sk_buff *skb, struct re_address *src, struc
                uh->check = CSUM_MANGLED_0;
 
        skb->protocol = htons(ETH_P_IP);
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,10,0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,4,78)
+       if (ip_route_me_harder(par->state->net, par->state->sk, skb, RTN_UNSPEC))
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(4,10,0)
        if (ip_route_me_harder(par->state->net, skb, RTN_UNSPEC))
 #elif LINUX_VERSION_CODE >= KERNEL_VERSION(4,4,0)
        if (ip_route_me_harder(par->net, skb, RTN_UNSPEC))
@@ -3447,7 +3449,9 @@ static int send_proxy_packet6(struct sk_buff *skb, struct re_address *src, struc
                uh->check = CSUM_MANGLED_0;
 
        skb->protocol = htons(ETH_P_IPV6);
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,10,0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,4,78)
+       if (ip6_route_me_harder(par->state->net, par->state->sk, skb))
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(4,10,0)
        if (ip6_route_me_harder(par->state->net, skb))
 #elif LINUX_VERSION_CODE >= KERNEL_VERSION(4,4,0)
        if (ip6_route_me_harder(par->net, skb))
```

**Comment by rfuchs** [comment]
The change seems to be between 5.9.8 and 5.9.9:
https://elixir.bootlin.com/linux/v5.9.8/source/include/linux/netfilter_ipv4.h
https://elixir.bootlin.com/linux/v5.9.9/source/include/linux/netfilter_ipv4.h
As well as 5.10-rc2 and 5.10-rc3:
https://elixir.bootlin.com/linux/v5.10-rc2/source/include/linux/netfilter_ipv4.h
https://elixir.bootlin.com/linux/v5.10-rc3/source/include/linux/netfilter_ipv4.h
Then for 5.4 there is:
https://elixir.bootlin.com/linux/v5.4.77/source/include/linux/netfilter_ipv4.h
https://elixir.bootlin.com/linux/v5.4.78/source/include/linux/netfilter_ipv4.h
I don't see it in any other 5.x branch yet. So there's a few special cases to be made...

**Comment by rfuchs** [comment]
Does 46a648890 work for you? (Untested)

**Comment by amessina** [comment]
@rfuchs I can confirm the build works on
* Fedora rawhide with kernel 5.10: https://download.copr.fedorainfracloud.org/results/amessina/mss/fedora-rawhide-x86_64/01783797-rtpengine-kmod/

**Comment by amessina** [comment]
I can also confirm the build patched with 46a64889071cb844550786096c251b804a234016 functions properly on 5.9.9-200.fc33.x86_64.

**Comment by ncopa** [comment]
I can confirm that it at least builds with 5.4.78.
